Source/WebCore/platform/glib/RunLoopObserverGLib.cpp

```diff
@@ -29,24 +29,33 @@
 #if USE(GLIB_EVENT_LOOP)
 namespace WebCore {
 
-void RunLoopObserver::schedule(PlatformRunLoop, OptionSet<Activity> activities)
+void RunLoopObserver::schedule(PlatformRunLoop runLoop, OptionSet<Activity> activities)
 {
-    if (m_runLoopObserver)
+    if (!runLoop)
+        runLoop = RunLoop::currentSingleton();
+    ASSERT(runLoop);
+
+    if (m_runLoopObserver) {
+        ASSERT(m_scheduledOnRunLoop == runLoop);
         return;
+    }
 
     m_runLoopObserver = ActivityObserver::create(static_cast<uint8_t>(m_order), activities, [this]() {
         runLoopObserverFired();
         return isRepeating() ? ActivityObserver::ContinueObservation::Yes : ActivityObserver::ContinueObservation::No;
     });
 
-    RunLoop::currentSingleton().observeActivity(*m_runLoopObserver);
+    m_scheduledOnRunLoop = runLoop;
+    m_scheduledOnRunLoop->observeActivity(*m_runLoopObserver);
```

Comment:
Since we always observe right after create and unobserve right before destroy, why don't we handle the run loop from ActivityObserver? We pass it to the constructor, so that the ActivityObserver is always attached to a run loop and it internally calls observe from constructor and unobserve from destructor.